lib/int_sqrt.c

```diff
@@ -16,25 +16,34 @@
  *
  * A very rough approximation to the sqrt() function.
  */
-unsigned long int_sqrt(unsigned long x)
+inline unsigned long int_sqrt(unsigned long x)
 {
-	unsigned long b, m, y = 0;
+	register unsigned long tmp;
+	register unsigned long place;
+	register unsigned long root = 0;
 
 	if (x <= 1)
 		return x;
 
-	m = 1UL << (__fls(x) & ~1UL);
-	while (m != 0) {
-		b = y + m;
-		y >>= 1;
+	place = 1UL << (BITS_PER_LONG - 2);
 
-		if (x >= b) {
-			x -= b;
-			y += m;
+	do{
+		place >>= 2;
+	}while(place > x);
+
+	do {
+		tmp = root + place;
+		root >>= 1;
+
+		if (x >= tmp)
+		{
+			x -= tmp;
+			root += place;
 		}
-		m >>= 2;
-	}
+		place >>= 2;
+	}while (place != 0);
 
-	return y;
+	return root;
 }
```

Comment:
I doubt that this implementation is faster. There are no real tests for this changes. For example, original implementation is heavily tested for performance: 083aa6a5d3b3248cb8078a8880fa036dd7a66d93 aab86217763b06bcb563cb69dbbff8d598b52a39  30493cc9dddb68066dcc4878015660fdaa8e0965 This patch looks really bad comparing to existing implementation.

Reply:
It might not on S10, on my J3 it did.